Fix the output when compiling plugins before a build starts

abertelrud opened the pull request:
Fixes the output for "Compile plugin" commands to use the progress tracker, just as for build commands.

### Motivation

This fixes potential output corruption in the console output.  No change in behaviour.

### Changes

- add delegate callbacks to the DefaultPluginScriptRunner so that clients can found out when compilation:
   1. will start
   2. did finish, or
   3. was skipped because the cache is already up-to-date
- change the printing of the output to use the same progress tracker as for build commands
- only show the direct compiler output in verbose mode, as for build commands

### Future improvements

An even better approach for a future PR would probably be to have the plugin compilation commands be registered as llbuild tasks, which would allow them to go through the regular dependency machinery.  This would still have to happen as a separate build step before the regular build, since prebuild commands etc have to run before the build and can contribute to target sources discovery etc (i.e. it can't be part of a regular build).

Meanwhile, this makes the output be like the others, and counts compiled plugins in the count of total and finished tasks.  I'll also extend this PR to do the same for running the plugins.

rdar://98367242

ktoso reviewed (commented):
Quite tricky after all actually huh. Thanks for looking into it!

abertelrud commented:
I'm not 100% happy with the fix here.  The actual compilation of the plugins use their own cache to avoid doing unneeded work, but llbuild doesn't know about that.  So while I've fixed the newline etc, the output for a null build is now:
```
[1/1] Compiling plugin MySourceGenBuildToolPlugin...
[2/2] Compiling plugin MySourceGenPrebuildPlugin...
Building for debugging...
Build complete! (0.16s)
```
(for an example case of two plugins) which is misleading, since the compilation is actually a no-op here.  We can fix that either by having llbuild handle the dependencies (more involved) or by emitting those lines from a callback from the plugin (less involved and probably what I'll do for the moment).

neonichu commented:
I think going the delegate route is probably preferable since letting llbuild handle the dependencies would likely mean that clients of libSwiftPM have to provide their own way of compiling plugins, provided we want to continue to keep the llbuild dependency separate.

abertelrud commented:
Yes, that's a good point (and was the original reason plugins were done outside of llbuild).  The code that does the compilation doesn't actually care how it is invoked, so we could factor it out so that the custom caching is present for clients that don't want to use llbuild, but that we trigger it from llbuild in the case of SwiftPM itself.  But as I'm describing that idea, it actually sounds unappealing to have differences in how the caching works.  So I think you're right.

abertelrud commented:
@swift-ci please smoke test

abertelrud commented:
Hm, I didn't change that part:
```
[1145/1175] Compiling Build BuildOperation.swift
/Users/ec2-user/jenkins/workspace/swift-package-manager-with-xcode-self-hosted-PR-osx/branch-main/swiftpm/Sources/Build/BuildOperation.swift:316:26: error: type of expression is ambiguous without more context
        let result = try tsc_await {
                     ~~~~^~~~~~~~~~~
```

abertelrud commented:
@swift-ci please smoke test

abertelrud commented:
> I think going the delegate route is probably preferable since letting llbuild handle the dependencies would likely mean that clients of libSwiftPM have to provide their own way of compiling plugins, provided we want to continue to keep the llbuild dependency separate.

Implemented the delegate route and updated the PR description.

tomerd commented on Sources/SPMBuildCore/PluginScriptRunner.swift:
```diff
@@ -61,18 +62,33 @@ public protocol PluginScriptRunner {
     func handleMessage(data: Data, responder: @escaping (Data) -> Void) throws
 }
-
+public extension PluginScriptRunnerDelegate {
```
does this extension need to be public?

abertelrud replied:
I thought it did, but I just tried now and it doesn't seem it does — thanks for calling that out.  I have fixed it.

abertelrud replied:
It does if we want to allow the delegate methods to be optional.  We could also require clients to implement them, thus getting rid of the extension.  Do you have a preference?

tomerd replied:
I generally prefer to make that an explicit decision of the client, this kind of extension will hide it from them

abertelrud replied:
The reason I added the defaults was because not all the callbacks make sense for all the use cases, e.g. if I chose to only compile the plugin, then having to implement the callbacks for running the plugin don't make sense.

That's probably an argument for splitting out the compiling and running into separate calls and separate delegate protocols.  I started down that road but it turned into a bit more of a change than I thought was warranted.  But maybe that's the right thing to do even in this PR (as opposed to a follow-on fix).

abertelrud replied:
OK:  I split this out into two separate protocols and use composition for the runner delegate, which makes things clearer.  I also got rid of the extensions providing default implementations, since SwiftPM has been moving away from those.

tomerd reviewed (approved):
seems right, @neonichu may have more informed opinion

abertelrud commented:
@swift-ci please smoke test

abertelrud commented:
@swift-ci please smoke test

abertelrud commented:
Still puzzled by this one, which would be ideal to figure out before we can merge it (though I guess isn't not in a required workflow):

```
/Users/ec2-user/jenkins/workspace/swift-package-manager-with-xcode-self-hosted-PR-osx/branch-main/swiftpm/Sources/Build/SPMSwiftDriverExecutor.swift:46:48: warning: 'resolveArgumentList(for:forceResponseFiles:quotePaths:)' is deprecated: use resolveArgumentList(for:,useResponseFiles:,quotePaths:)
        let arguments: [String] = try resolver.resolveArgumentList(for: job,
                                               ^
/Users/ec2-user/jenkins/workspace/swift-package-manager-with-xcode-self-hosted-PR-osx/branch-main/swiftpm/Sources/Build/SPMSwiftDriverExecutor.swift:76:53: warning: 'resolveArgumentList(for:forceResponseFiles:quotePaths:)' is deprecated: use resolveArgumentList(for:,useResponseFiles:,quotePaths:)
        let (args, usedResponseFile) = try resolver.resolveArgumentList(for: job,
                                                    ^
[1146/1175] Compiling Build BuildOperation.swift
/Users/ec2-user/jenkins/workspace/swift-package-manager-with-xcode-self-hosted-PR-osx/branch-main/swiftpm/Sources/Build/BuildOperation.swift:316:26: error: type of expression is ambiguous without more context
        let result = try tsc_await {
                     ~~~~^~~~~~~~~~~
[1147/1175] Compiling Build LLBuildManifestBuilder.swift
/Users/ec2-user/jenkins/workspace/swift-package-manager-with-xcode-self-hosted-PR-osx/branch-main/swiftpm/Sources/Build/LLBuildManifestBuilder.swift:239:76: warning: 'makeDiagnosticsEngine()' is deprecated: temporary for transition DiagnosticsEngine -> DiagnosticsEmitter
                                diagnosticsEngine: self.observabilityScope.makeDiagnosticsEngine(),
                                                                           ^
```

@neonichu any ideas?

neonichu commented:
My guess would be that the older compiler used by the self-hosted workflow is the problem?

abertelrud commented:
> My guess would be that the older compiler used by the self-hosted workflow is the problem?

Thank you, that makes a lot of sense!  I'll try with an Xcode 13.x compiler and add whatever disambiguation is needed to make sure it keeps compiling there.

abertelrud commented:
@swift-ci please smoke test
